### DISCUSSION TITLE
Working QML client

### DISCUSSION THREAD

**Discussion by danryu** [jamulussoftware/jamulus#3454] [General]
Hi all,
While setting myself the challenge to learn QML, I have succeeded in rewriting the core Jamulus client GUI with QML, removing all widget code in the process.
To put it mildly, it was a large, non-trivial effort to disentangle all the widget display logic from the business logic.

All existing business logic is largely preserved, with small changes and additions here and there.
In general, things are much simplified when it comes to UI element handling. No need, for example, for the clever audiomixerboard template stuff, or creating and hiding 250 channelfader objects :)

**Key Changes**
- clientdlg -> removed, business logic moved to client, display moved to QML
- settingsdlg -> removed, business logic moved to settings, display moved to QML
- chatdlg -> converted to chatbox
- audiomixerboard -> CChannelFader and CAudioMixerBoard classes refactored as necessary
- all *.qml files representing key components in the UI

**Design**
The UI design is just a demo to contain all the functioning elements. It can easily be adapted as necessary.
For this reason it has almost no styling applied.

**Features**
All the core features of the client have been reproduced in the QML version. There may still be some subtle bugs, but it is already quite usable (only tested on Windows 11). 

**Exclusions**
To reduce the scope of the project (already large enough) I have not implemented the following, but they should not be challenging to re-introduce:
- server chooser dialog
- instruments/skills/flags
- reverb
- what's this stuff
- translations / language chooser stuff

I invite you to build and test it locally, and offer any suggestions or enhancements.
Especially on macOS as I don't have any Apple hardware at the moment.

https://github.com/danryu/jamulus/tree/qmldev2
So far, tested and working on:
- Qt 6.8.1 - Windows 11 (MSVC2022)
- Qt 6.8.1 - Linux

![Screenshot 2024-12-29 144841](https://github.com/user-attachments/assets/9575f3b8-f6cc-4ec4-b7bc-b18788cda7b4)
![Screenshot 2024-12-29 144919](https://github.com/user-attachments/assets/9cbb7df2-6773-45e6-9d6f-edada7394787)

**Comment by ann0see** [comment]
Wow. Thanks for this huge effort! I think that's exactly what Jamulus 4 should do. Can I ask you to open a PR?
This PR will most likely serve as draft and discussion only. But I'd strongly support this move.

**Reply by ann0see** [reply]
I think this is also something where we'd need reviews from multiple people. In the best case also Volker.

**Reply by danryu** [reply]
Ok I went ahead and opened https://github.com/jamulussoftware/jamulus/pull/3457.
Hopefully it will prove useful ...

**Reply by ann0see** [reply]
Thanks.

**Comment by ann0see** [comment]
> The UI design is just a demo to contain all the functioning elements. It can easily be adapted as necessary.

I think the end goal would be to mirror the current GUI quite closely to not upset users. Of course there will be some visual changes.

**Reply by danryu** [reply]
Sure. One good thing about the QML implementation is that it's relatively trivial to move the visual elements around as you want.

**Comment by rdica** [comment]
I attempted to build the fork/branch on Ubuntu Studio 22.04 using QT 5.15.3 and the build fails, here's the last couple lines of the compile attempt using `qmake "CONFIG+=noupcasename" Jamulus.pro && make clean && make -j8`:

```
g++ -c -pipe -O2 -std=gnu++1z -D_REENTRANT -Wall -Wextra -fPIC -DAPP_VERSION=\"3.12.0qml\" -DCUSTOM_MODES -D_REENTRANT -DHAVE_LRINTF -DHAVE_STDINT_H -DWITH_JACK -DOPUS_X86_MAY_HAVE_SSE -DOPUS_X86_MAY_HAVE_SSE2 -DOPUS_X86_MAY_HAVE_SSE4_1 -DOPUS_X86_PRESUME_SSE=1 -DOPUS_X86_PRESUME_SSE2=1 -DCPU_INFO_BY_C -DOPUS_BUILD=1 -DUSE_ALLOCA=1 -DOPUS_HAVE_RTCD=1 -DHAVE_LRINTF=1 -DHAVE_LRINT=1 -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_QUICKCONTROLS2_LIB -DQT_QUICK_LIB -DQT_GUI_LIB -DQT_QMLMODELS_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_CONCURRENT_LIB -DQT_CORE_LIB -I. -Isrc -Ilibs/opus/include -Ilibs/opus/celt -Ilibs/opus/silk -Ilibs/opus/silk/float -Ilibs/opus/silk/fixed -Ilibs/opus -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtSvg -I/usr/include/x86_64-linux-gnu/qt5/QtWidgets -I/usr/include/x86_64-linux-gnu/qt5/QtQuickControls2 -I/usr/include/x86_64-linux-gnu/qt5/QtQuick -I/usr/include/x86_64-linux-gnu/qt5/QtGui -I/usr/include/x86_64-linux-gnu/qt5/QtQmlModels -I/usr/include/x86_64-linux-gnu/qt5/QtQml -I/usr/include/x86_64-linux-gnu/qt5/QtNetwork -I/usr/include/x86_64-linux-gnu/qt5/QtXml -I/usr/include/x86_64-linux-gnu/qt5/QtConcurrent -I/usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o main.o src/main.cpp
g++ -c -pipe -O2 -std=gnu++1z -D_REENTRANT -Wall -Wextra -fPIC -DAPP_VERSION=\"3.12.0qml\" -DCUSTOM_MODES -D_REENTRANT -DHAVE_LRINTF -DHAVE_STDINT_H -DWITH_JACK -DOPUS_X86_MAY_HAVE_SSE -DOPUS_X86_MAY_HAVE_SSE2 -DOPUS_X86_MAY_HAVE_SSE4_1 -DOPUS_X86_PRESUME_SSE=1 -DOPUS_X86_PRESUME_SSE2=1 -DCPU_INFO_BY_C -DOPUS_BUILD=1 -DUSE_ALLOCA=1 -DOPUS_HAVE_RTCD=1 -DHAVE_LRINTF=1 -DHAVE_LRINT=1 -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_QUICKCONTROLS2_LIB -DQT_QUICK_LIB -DQT_GUI_LIB -DQT_QMLMODELS_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_CONCURRENT_LIB -DQT_CORE_LIB -I. -Isrc -Ilibs/opus/include -Ilibs/opus/celt -Ilibs/opus/silk -Ilibs/opus/silk/float -Ilibs/opus/silk/fixed -Ilibs/opus -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtSvg -I/usr/include/x86_64-linux-gnu/qt5/QtWidgets -I/usr/include/x86_64-linux-gnu/qt5/QtQuickControls2 -I/usr/include/x86_64-linux-gnu/qt5/QtQuick -I/usr/include/x86_64-linux-gnu/qt5/QtGui -I/usr/include/x86_64-linux-gnu/qt5/QtQmlModels -I/usr/include/x86_64-linux-gnu/qt5/QtQml -I/usr/include/x86_64-linux-gnu/qt5/QtNetwork -I/usr/include/x86_64-linux-gnu/qt5/QtXml -I/usr/include/x86_64-linux-gnu/qt5/QtConcurrent -I/usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o protocol.o src/protocol.cpp
In file included from src/main.cpp:36:
src/settings.h:32:14: fatal error: QGUIApplication: No such file or directory
   32 | #    include <QGUIApplication>
      |              ^~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:2310: main.o] Error 1
make: *** Waiting for unfinished jobs....
```

**Reply by danryu** [reply]
Not sure why your build is failing to find QGUIApplication, it should be included with QT+=gui (https://doc.qt.io/qt-6/qguiapplication.html).

However, you're attempting to build with Qt 5, which is dead and unsupported for 18 months now, unless you're a subscription license holder, in which case it will be dead and unsupported in May 2025.
Best to switch to a LTS Qt 6 release like the latest, 6.8.

**Reply by dingodoppelt** [reply]
Hi and thanks for your efforts! Very much appreciated.
Sadly I have the same problem. Adding gui to Jamulus.pro didn't work either. QT 6.8.1 was used.

**Reply by rdica** [reply]
@danryu The linux distro ships with QT 5.15.3, the desktop env is built on it. QT6 packages are available but I have to check if I can install both versions simultaneously without causing issues for the desktop and existing QT apps.

I was able to get a little further in the build by altering `src/settings.h` and changing:
```
diff --git a/src/settings.h b/src/settings.h
index 0c86bec4..b9167193 100644
--- a/src/settings.h
+++ b/src/settings.h
@@ -29,7 +29,8 @@
 #include <QSettings>
 #include <QDir>
 #ifndef HEADLESS
-#    include <QGUIApplication>
+// #    include <QGUIApplication>
+#    include <QtGui/qguiapplication.h>
 #endif
 #include "global.h"
 // #ifndef SERVER_ONLY
```

Build errors out on `src/audiomixerboard.cpp`. I don't include the errors unless you want them.

I will investigate QT6 install and if successful will retry build.

**Reply by dingodoppelt** [reply]
@rdica Thanks! That patch solved it for me. You might want to try qt6 yourself ;)
@danryu Impressive work! Looks great already :+1:

**Reply by rdica** [reply]
@dingodoppelt Nice! Can you detail your build env, and do you have both QT5 **and** 6 installed?

**Reply by dingodoppelt** [reply]
@rdica Yes, I do have both qt5 and 6 installed. I simply used qmake6 instead of qmake, everything else kept the same. I'm on arch with kde.

**Reply by ann0see** [reply]
@rdica it might make sense to open a PR to @danryu 's repo to this branch so that he can merge it.

**Reply by danryu** [reply]
Bit of a confusing situation! 
main.cpp already does
`#   include <QGuiApplication>`
with lower-case "ui", so I've updated the settings.h include to be the same. This is after all what the official docs state: https://doc.qt.io/qt-6/qguiapplication.html.
(Perhaps this is an odd difference in behavior between the same Qt versions on Windows and Linux?)

I also removed a redundant include of qguiapplication.h in audiomixerboard.h.

So in theory the build should work now with Qt6 on all platforms.

**Reply by dingodoppelt** [reply]
> So in theory the build should work now with Qt6 on all platforms.

It does work for me :+1:

**Reply by rdica** [reply]
I was able to build after lowercasing `ui` in the include.

> (Perhaps this is an odd difference in behavior between the same Qt versions on Windows and Linux?)

Given that Windows is case-insensitive, uppercasing "QGUI" wouldn't have caused an issue is my guess. Mac and Linux are case-sensitive. Good catch!

**Reply by danryu** [reply]
Thanks for the original catch :) I only use Windows when necessary so I forget these details sometimes. My brain remembers that NTFS is case-sensitive, but forgets that the OS ignores the case anyway

**Reply by rdica** [reply]
I'm experiencing inconsistent connection issues which I haven't yet been able to lockdown. At first I couldn't connect to any IP or hostname, but when I added `-6` option I was able to connect to IPv6 hostnames and IPs. I was then able to connect to IPv4 IPs as well, but not everytime. I could see outgoing traffic from client while running tcpdump on client machine but no traffic from the src hitting the destination server. No errors showing in the terminal I started the client from either.
Curiouser and curiouser.
I'll keep testing and see if I can pinpoint the behaviour.

**Reply by ann0see** [reply]
This could stem from missing NAT traversal traffic since the client probably doesn't talk to a directory yet.

**Reply by rdica** [reply]
Ok, I found the issue. The Jamulus protocol has a NAT/firewall hole punch feature that gets intiated when you select a server from the connect dialog. Due to the lack of a directory listing in this QML version, you can only direct connect to servers. Some cloud/hosting service firewalls don't permit direct connections, and those are ones I couldn't connect to. I'd forgotten this in my initial testing yesterday. It explains the inconsistency I was experiencing on with both IPv6 and IPv4 servers.

@danryu There are client behaviours I'm seeing. Would you like me to detail them, and if yes, should we do that elsewhere. I don't think we should highjack this discussion thread with minutiae.

**Reply by danryu** [reply]
Thanks @rdica, that would be great. I've opened Issues on my fork - perhaps best to log problems there, if you wish?

> Some cloud/hosting service firewalls don't permit direct connections

I'd forgotten about this behavior, as I was only testing direct connections to get the UI working. Could you maybe remind me of the use case here? Is this for servers being run off home connections primarily? I'm just having a dig now to see where and how this NAT traversal is implemented.

**Reply by ann0see** [reply]
> I'm just having a dig now to see where and how this NAT traversal is implemented.

I think this will only work if the connection dialog is implemented. So I'd say it's for a second iteration in future.

**Reply by rdica** [reply]
@danryu Thanks! I've submitted three issues I came across. I did not mention my experience with connecting and segfaulting as thats been covered below and I too cannot reproduce it consistently.

**Reply by danryu** [reply]
Thanks for tracking those. I've made an issue for the segfault issue, which I'll hopefully narrow down while looking at the settings save/load process. Meanwhile, any other issue-catches welcome!

**Comment by ann0see** [comment]
macOS builds without issue.
Dark mode is broken however
![Jamulus QML build broken darkmode](https://github.com/user-attachments/assets/47ddc565-32d1-46de-aeb5-d9d5d1e821b8)

If we're lucky, the following .zip could contain a build that works on Intel Macs:
[Jamulus_QML.zip](https://github.com/user-attachments/files/18286721/Jamulus_QML.zip)

**Reply by ann0see** [reply]
Also settings look a bit off

![Settings QML](https://github.com/user-attachments/assets/90ac1270-7a15-4a10-b741-e10995a451d4)

**Reply by ann0see** [reply]
I got it to crash with a segfault while trying to connect to a server. Not reproducible.
```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               Jamulus [1676]
Path:                  /Users/USER/Downloads/*/Jamulus.app/Contents/MacOS/Jamulus
Identifier:            live.jamulus.Jamulus
Version:               3.12 (3.12.0qml)
Code Type:             X86-64 (Native)
Parent Process:        launchd [1]
User ID:               502

Date/Time:             2025-01-01 22:38:00.0232 +0100
OS Version:            macOS 14.6.1 (23G93)
Report Version:        12
Anonymous UUID:        AA466EE5-29A9-7BAE-634A-FF847A7DDAB2


Time Awake Since Boot: 680 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes:       0x0000000000000001, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [1676]

VM Region Info: 0 is not in any region.  Bytes before following region: 4509360128
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      10cc76000-10cdc5000    [ 1340K] r-x/r-x SM=COW  /Users/USER/Downloads/*/Jamulus.app/Contents/MacOS/Jamulus

Kernel Triage:
VM - (arg = 0x3) mach_vm_allocate_kernel failed within call to vm_map_enter
VM - (arg = 0x3) mach_vm_allocate_kernel failed within call to vm_map_enter
VM - (arg = 0x3) mach_vm_allocate_kernel failed within call to vm_map_enter


Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   Jamulus                       	       0x10cd1423e CAudioMixerBoard::SetRemoteFaderIsMute(int, bool) + 14
1   QtCore                        	       0x1101d7bba QtPrivate::QSlotObjectBase::call(QObject*, void**) + 25 (qobjectdefs_impl.h:486) [inlined]
2   QtCore                        	       0x1101d7bba void doActivate<false>(QObject*, int, void**) + 1594 (qobject.cpp:4124)
3   Jamulus                       	       0x10cd73919 CChannel::MuteStateHasChangedReceived(int, bool) + 73
4   QtCore                        	       0x1101d7bba QtPrivate::QSlotObjectBase::call(QObject*, void**) + 25 (qobjectdefs_impl.h:486) [inlined]
5   QtCore                        	       0x1101d7bba void doActivate<false>(QObject*, int, void**) + 1594 (qobject.cpp:4124)
6   Jamulus                       	       0x10cd75fc9 CProtocol::MuteStateHasChangedReceived(int, bool) + 73
7   Jamulus                       	       0x10cc9cdb4 CProtocol::EvaluateMuteStateHasChangedMes(CVector<unsigned char> const&) + 36
8   Jamulus                       	       0x10cc9c8fd CProtocol::ParseMessageBody(CVector<unsigned char> const&, int, int) + 1005
9   Jamulus                       	       0x10ccf526c QtPrivate::FunctorCall<QtPrivate::IndexesList<0, 1, 2, 3>, QtPrivate::List<int, int, CVector<unsigned char>, CHostAddress>, void, void (CChannel::*)(int, int, CVector<unsigned char>, CHostAddress)>::call(void (CChannel::*)(int, int, CVector<unsigned char>, CHostAddress), CChannel*, void**)::'lambda'()::operator()() const + 204
```

Seems like a null pointer de reference to me:

```
  "vmRegionInfo" : "0 is not in any region.  Bytes before following region: 4509360128\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      10cc76000-10cdc5000    [ 1340K] r-x\/r-x SM=COW  \/Users\/USER\/Downloads\/*\/Jamulus.app\/Contents\/MacOS\/Jamulus",
  "exception" : {"codes":"0x0000000000000001, 0x0000000000000000","rawCodes":[1,0],"type":"EXC_BAD_ACCESS","signal":"SIGSEGV","subtype":"KERN_INVALID_ADDRESS at 0x0000000000000000"},
  "termination" : {"flags":0,"code":11,"namespace":"SIGNAL","indicator":"Segmentation fault: 11","byProc":"exc handler","byPid":1676},
  "ktriageinfo" : "VM - (arg = 0x3) mach_vm_allocate_kernel failed within call to vm_map_enter\nVM - (arg = 0x3) mach_vm_allocate_kernel failed within call to vm_map_enter\nVM - (arg = 0x3) mach_vm_allocate_kernel failed within call to vm_map_enter\n",
  "vmregioninfo" : "0 is not in any region.  Bytes before following region: 4509360128\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      10cc76000-10cdc5000    [ 1340K] r-x\/r-x SM=COW  \/Users\/USER\/Downloads\/*\/Jamulus.app\/Contents\/MacOS\/Jamulus",
  "extMods" : {"caller":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"system":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"targeted":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"warnings":0},
  "faultingThread" : 0,
  "threads" : [{"triggered":true,"id":13761,"instructionState":{"instructionStream":{"bytes":[208,137,10,72,139,136,160,0,0,0,66,131,60,177,0,72,139,77,200,15,149,1,72,139,136,184,0,0,0,66,131,60,177,0,72,139,77,192,15,149,1,72,139,128,208,0,0,0,66,139,4,176,72,139,77,16,137,1,176,1,72,131,196,24,91,65,92,65,93,65,94,65,95,93,195,15,31,128,0,0,0,0,85,72,137,229,133,246,120,30,137,240,72,139,79,32,72,139,60,193,128,127,34,0,15,149,192,56,208,116,9,136,87,34,93,233,58,216,6,0,93,195,15,31,132,0,0,0,0,0,85,72,137,229,65,87,65,86,65,85,65,84,83,80,65,137,207,137,243,73,137,254,76,99,230,69,133,192,117,23,15,182,210,76,137,247,137,222,242,15,17,77,208,232,112,222,6,0,242,15,16,77,208,77,139,110,32,75,139,68,229,0],"offset":96}},"threadState":{"r13":{"value":140701839089920},"rax":{"value":0},"rflags":{"value":66118},"cpu":{"value":2},"r14":
 ```

**Reply by rdica** [reply]
I experienced this too, mostly when I resized the window before trying to connect, haven't seen a segfault with starting window size (seems the resize isn't saved upon quit).

**Reply by danryu** [reply]
Not sure why this is intermittent, but I didn't do anything with SetRemoteFaderIsMute yet (just a little bit out of scope :) so there's probably something to tidy up there.

Also I don't remember doing window geometry saving specifically, so that makes sense.

**Reply by danryu** [reply]
> macOS builds without issue. Dark mode is broken however 

Yes I think this is where Qt Quick Controls comes in https://doc.qt.io/qt-6/qtquickcontrols-styles.html. There's a bunch of style options and platform-specific settings that I haven't really looked into yet.

**Comment by ann0see** [comment]
Feedback detection "works", but the constructed message cannot be closed...
![Feedback detection](https://github.com/user-attachments/assets/4936633d-760c-4b7e-8794-eeca9301a31b)

**Reply by danryu** [reply]
I didn't put a "close" or "X" or somesuch yet, but the popup is closable by pressing ESC or clicking anywhere outside the popup.
